metadata-ingestion/src/datahub/api/entities/dataproduct/dataproduct.py

```diff
@@ -0,0 +1,525 @@
+            )
+            domain_urn = domain_registry.get_domain_urn(parsed_data_product.domain)
+            parsed_data_product._resolved_domain_urn = domain_urn
+            return parsed_data_product
+
+    @classmethod
+    def from_datahub(cls, graph: DataHubGraph, id: str) -> "DataProduct":
+        # data_product_dict = graph.get_entity_raw(id)
+        data_product_properties: Optional[
+            DataProductPropertiesClass
+        ] = graph.get_aspect(id, DataProductPropertiesClass)
+        domains: Optional[DomainsClass] = graph.get_aspect(id, DomainsClass)
+        assert domains, "Data Product must have an associated domain. Found none."
+        owners: Optional[OwnershipClass] = graph.get_aspect(id, OwnershipClass)
+        yaml_owners: Optional[List[Union[str, DataProduct.Ownership]]] = None
+        if owners:
+            yaml_owners = []
+            for o in owners.owners:
+                if o.type == OwnershipTypeClass.TECHNICAL_OWNER:
+                    yaml_owners.append(o.owner)
+                else:
+                    yaml_owners.append(
+                        DataProduct.Ownership(id=o.owner, type=str(o.type))
+                    )
+        glossary_terms: Optional[GlossaryTermsClass] = graph.get_aspect(
+            id, GlossaryTermsClass
+        )
+        tags: Optional[GlobalTagsClass] = graph.get_aspect(id, GlobalTagsClass)
+        return DataProduct(
+            id=id,
+            display_name=data_product_properties.name
+            if data_product_properties
+            else None,
+            domain=domains.domains[0],
+            description=data_product_properties.description
+            if data_product_properties
+            else None,
+            assets=[e.destinationUrn for e in data_product_properties.assets or []]
+            if data_product_properties
+            else None,
+            owners=yaml_owners,
+            terms=[term.urn for term in glossary_terms.terms]
+            if glossary_terms
+            else None,
+            tags=[tag.tag for tag in tags.tags] if tags else None,
+            properties=data_product_properties.customProperties
+            if data_product_properties
+            else None,
+            external_url=data_product_properties.externalUrl
+            if data_product_properties
+            else None,
+        )
+
+    def _patch_ownership(
+        self,
+        original_owners: Optional[List[Union[str, DataProduct.Ownership]]],
+        original_ownership_list: Optional[List[Any]],
+    ) -> Tuple[bool, Optional[List[Any]]]:
+        new_owner_type_map = {}
+        for new_owner in self.owners or []:
+            if isinstance(new_owner, DataProduct.Ownership):
+                new_owner_type_map[new_owner.id] = new_owner.type
+            else:
+                new_owner_type_map[new_owner] = "TECHNICAL_OWNER"
+        owners_matched = set()
+        patches: Dict[str, Union[dict, list]] = {"ADD": [], "DROP": {}, "REPLACE": {}}
```

Comment:
what if we broke this into 3 different variables - I think the resulting code would be much more readable and wouldn't need all of the type ignores